systemschema: Added tests that ensure consistent initial system tables

Xiang-Gu opened the pull request:
Previously, there is no tests that checks the validity/sanity of the system tables after bootstrapping/upgrading (recall those table defs are hard-coded).
This PR attempts to at least have some tests to help us gain more confidence
that nothing about system tables were messed up after bootstrapping/upgrading.
It simply use USE system; SHOW CREATE ALL TABLES; during those two situations (bootstrapping and upgrading) and compare the output with expected, predefined output.

Namely, the following two packages are modified:

1. pkg/sql/systemschema_test/: this is a newly added package. We added a data-driven that spins up a test cluster and validate system tables are initialized as expected.

2. pkg/cmd/roachtest/tests/: added a roachtest that spins up a cluster with
older binary release's version, upgrades it, and ensure the system
tables are consistent.


Fixes: #72462

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/77970)
<!-- Reviewable:end -->

Xiang-Gu committed:
systemschema: Added tests that ensure consistent initial system tables

Previously, there is no tests that checks the validity/sanity of the
system tables after bootstrapping/upgrading (recall those table defs
are hard-coded).
This PR attempts to at least have some tests to help us gain more
confidence that nothing about system tables were messed up after
bootstrapping/upgrading.
It simply use USE system; SHOW CREATE ALL TABLES; during those two
situations (bootstrapping and upgrading) and compare the output with
expected, predefined output.

Namely, the following two packages are modified:

1. pkg/sql/systemschema_test/: this is a newly added package. We added
a data-driven that spins up a test cluster and validate system tables
are initialized as expected.

2. pkg/cmd/roachtest/tests/: added a roachtest that spins up a cluster
with older binary release's version, upgrades it, and ensure the system
tables are consistent.

Fixes: #72462

Release note: None

ajwerner reviewed (commented):
Do we need to check in this testdata? It seems like it's going to be brittle when we add new tables. Can we just boot the different version, pull what we expect and then do the upgrade, see what we have, and then compare them?

> Reviewed 3 of 9 files at r1, 1 of 4 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77970#-:-MzW4O2E41mctBA7CP2F:bi428mq)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)



<!-- Sent from Reviewable.io -->

Xiang-Gu reviewed (commented):
> Do we need to check in this testdata? It seems like it's going to be brittle when we add new tables. Can we just boot the different version, pull what we expect and then do the upgrade, see what we have, and then compare them?

Can you elaborate it further?  We currently have two test data in this PR — one for the bootstrapping test and the other for the upgrading test. Are you suggesting that for the upgrading test, we remove its testdata file, boot a cluster on an old version, issue show create table query to retrieve the current system table definition `d1`, upgrade the cluster, issue the query again to retrieve table definitions `d2` after migration. But then what are we comparing against? We need to somehow compare `d2` against the expected output, which is now the testdata file.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77970#-:-MzWC_eO95jtcN2-dNX1:b-nywpug)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)



<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
The bootstrap one is good. 

The one I'm talking about is `validate_system_schema_after_version_upgrade_expected_output`, for that, could we start by booting the cluster in the new version, run the query, then wipe that cluster, boot it in the predecessor version, upgrade, run the query, and compare. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77970#-:-MzWGd-P2n_gSRSllC-3:bilcxyk)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)



<!-- Sent from Reviewable.io -->

Xiang-Gu reviewed (commented):
Yeah, that's a good idea. I've modified the roachtest part to do this. 

RFAL

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77970#-:-MzWRzUI95hFkqqvd4Fx:blc98ug)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)



<!-- Sent from Reviewable.io -->

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>

> Reviewed 2 of 9 files at r1, 1 of 4 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77970#-:-MzXj40r75ebxcYwPvDY:b-jbsf8z)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @Xiang-Gu)



<!-- Sent from Reviewable.io -->

Xiang-Gu commented:
bors r+
